introduce a max_connection_ids transport parameter (#1998)
quicwg/base-drafts: draft-ietf-quic-transport.md

the connection ID via a RETIRE_CONNECTION_ID frame
({{frame-retire-connection-id}}).

An endpoint SHOULD ensure that its peer has a sufficient number of available and
unused connection IDs. Endpoints store received connection IDs for future use. 

> They advertise the limit how many unretired connection IDs they are willing to 

Reviewer: Word missing?

Author: Where?

Reviewer 2: "advertise the limit how many" could probably do with a limit "on" how many, but it might read better to simply "advertise how many".

Author: Thanks for the suggestion, @MikeBishop!

> the transport parameters. An endpoint MUST NOT provide more connection IDs than

Reviewer 3: Suggestion to replace this previous sentence: "An endpoint uses a transport parameter to advertise the maximum number of connection IDs it can store for future use."

this limit. An endpoint MUST treat receipt of more than this number of

> connection IDs as an error of type CONNECTION_ID_LIMIT_ERROR. If an endpoint has

Reviewer 3: Why is this error necessary? A peer can simply drop the ones that it can't store.

Author: I regard this as an equivalent to flow control and stream ID flow control. We have errors there if the peer violates a limit. It feels weird to me to set a limit, but then to silently tolerate a violation of that limit. And it will be a lot easier to develop compliant implementation if we actually enforce the limit.

provided its peer with the maximum number of connection IDs, the earliest time

> it can provide a new connection ID is when the peer retires a connection ID.

Reviewer 3: suggestion: "... connection IDs, it can only provide a new connection ID when the peer retires one."

the connection ID via a RETIRE_CONNECTION_ID frame
({{frame-retire-connection-id}}).

An endpoint SHOULD ensure that its peer has a sufficient number of available and
unused connection IDs. Endpoints store received connection IDs for future use.
An endpoint uses a transport parameter to advertise the maximum number of
connection IDs it can store for future use. An endpoint MUST NOT provide more
connection IDs than this limit. An endpoint MUST treat receipt of more than this

> number of connection IDs as an error of type CONNECTION_ID_LIMIT_ERROR. If an

Reviewer 3: Do we want a new error for this, or can we live with PROTOCOL_VIOLATION?

Author: That's the old discussion of error codes vs. error messages, right? Considering the fact that we have separate *_LIMIT_ERRORs for flow control and stream IDs, I thought it would be consistent to have an error code for this as well. I could live with PROTOCOL_VIOLATION as well, if people are unhappy with that.

the connection ID via a RETIRE_CONNECTION_ID frame
({{frame-retire-connection-id}}).

An endpoint SHOULD ensure that its peer has a sufficient number of available and
unused connection IDs. Endpoints store received connection IDs for future use.
An endpoint uses a transport parameter to advertise the number of connection IDs
it can store for future use. An endpoint SHOULD NOT provide more connection IDs
than this limit. If an endpoint has provided its peer with the maximum number of
connection IDs, it SHOULD only provide a new connection ID when the peer retires
one. Endpoints that initiate migration and require non-zero-length connection
IDs SHOULD provide their peers with new connection IDs before migration, or risk

> the peer closing the connection.

Reviewer 4: There are a lot of SHOULDs here, but it's a bit light on the associated justification.  I think that we'll need some more text about why an endpoint might choose not to follow this advice.  Having everything in the one paragraph will make that difficult.

the connection ID via a RETIRE_CONNECTION_ID frame
({{frame-retire-connection-id}}).

An endpoint SHOULD ensure that its peer has a sufficient number of available and
unused connection IDs. Endpoints store received connection IDs for future use.

> An endpoint uses a transport parameter to advertise the number of connection IDs

Reviewer 5: ```suggestion
An endpoint uses the active_connection_id_limit transport parameter to advertise the number of connection IDs
```

it can store for future use. An endpoint SHOULD NOT provide more connection IDs

> than this limit. If an endpoint has provided its peer with the maximum number of

Reviewer 3: ```suggestion
than the peer's limit. If an endpoint has provided its peer with the maximum number of
```

      max_ack_delay(11),
      disable_migration(12),
      preferred_address(13),

>       max_connection_ids(14),

Author: Do we need to renumber these again?

a preferred address.  A server MUST treat receipt of any of these transport
parameters as a connection error of type TRANSPORT_PARAMETER_ERROR.

max_connection_ids (0x000e):

: The maximum number of connection IDs that the peer is willing to store.

>   This value includes only connection IDs sent in NEW_CONNECTION_ID frames.

Reviewer 3: The maximum number of connection IDs that this endpoint can store for future use. These connection IDs are issued by the peer and sent in NEW_CONNECTION_ID frames.

Author: I *can* store a lot. But I might not be willing to.

Reviewer 3: You have the two sides mixed up. This param should convey the max that _this_ endpoint is willing to store. For clarity, I also suggest that you explicitly call out who is sending the new connection IDs: "... includes only connections IDs sent by the peer..."

Reviewer 6: I think what @janaiyengar is pointing out is that "the _peer_ is willing to store" needs to be changed to "the _endpoint_ is willing to store".

Reviewer 7: Isn't the number of stored connection IDs symmetric? In practice, each address change requires both sides to change their CIDs (except for zero-length cases).

I'm not sure I see any point in ever storing more than min(my_max_connection_ids, your_max_connection_ids)

Reviewer 4: The need for reciprocity in most cases that consume a connection ID does tend to lead to the numbers being approximately the same.  But it's not symmetric in the sense that they need to be equal.  The design is therefore asymmetric.  

If you consider a client that is moving between networks of questionable quality, they can burn connection IDs on futile probes, leading to greater consumption of connection IDs.  That suggests that a client might store more spares than a server.

>   This value includes only connection IDs sent in NEW_CONNECTION_ID frames.

Reviewer 2: Why "only" connection IDs sent in NCID frames?  The CIDs provided in the handshake and SPA consume space in this list, get retired, etc. If they don't count toward this limit, that would suggest that I'll be temporarily 1-2 over the limit early in the connection, then drift down to the limit as I retire those CIDs. That doesn't seem like the desired effect.

Reviewer 2: Hm, that's true.  But the reverse creates an odd situation where the server can issue 7 CIDs when the client has said they only want 5, but the client can't (or shouldn't) reduce its value by 2 and constrain itself for the lifetime of the connection because of a start-only overflow.

Perhaps the server SHOULD NOT issue more total CIDs than this, but MUST NOT issue more CIDs via NCID?

Reviewer 2: Worst case, from an implementation perspective, if you hit your limit and still have the handshake CIDs, retire them and the right things will happen.

Reviewer 3: Ugh. I don't like including the CID with SPA in this count for the reason @ianswett mentions. Given that the only CID that a client will _store_ is the one that is sent on an SPA, and even that will get used immediately (or get discarded), I think it's reasonable to say that the number of _stored_ CIDs is limited by max_connection_ids.

Reviewer 8: SPA=?

Reviewer 6: I think this discussion is a good reason to not have the enforcement (i.e. close the connection if peer exceeds the maximum).

As @ekr has pointed out in https://github.com/quicwg/base-drafts/issues/1994#issuecomment-441305154, the `max_connection_ids` can be a purely advisory value without defining any error conditions.

Reviewer 6: @dtikhonov Server Preferred Address

Author: I'm not sure why enforcing this creates pressure to increase the max. Sure, you need to advertise a value that actually makes sense (i.e. you should request as many CIDs as you think you'll need), and not rely on the peer overflowing this limit (which is exactly what we're trying to avoid in the first place).

Reviewer 3: @kazuho: I think that's fair, and I'm ok with not closing the connection if a peer violates my max_connection_id. We need to be clear that an endpoint that drops a NEW_CID because it has no space MUST NOT send a RETIRE_CID. This is an easy bug, so it's worth noting.

Reviewer 9: If a server has sent a CID as part of SPA, what sequence number does it use for the next NCID?

Author: @huitema I think this is unrelated to this PR. We have text about that in https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#issue-cid.

Reviewer 9: It is *not* unrelated. The sequence number is a direct statement on the number of CID out there. I would very much like to have just one number to test. Say, state "sequence number should be lower than the number of retired CID plus the max number of CID accepted by the peer".

Reviewer 5: I think the restriction you're suggesting is a nice simplification Christian.

I'd also like to close the connection if the limit is exceeded, but I know there was some opposition to that in Tokyo that I didn't fully understand.

Reviewer 2: Iff you count the handshake and SPA CIDs, an overflow just from transport parameters is possible if set to 0/1, as @marten-seemann pointed out earlier.  Overflows are also possible based on the advice to replace upon seeing a new CID enter use, though I think that's best addressed by just not recommending/allowing that.

Reviewer 9: @MikeBishop yes indeed, the server always sends at least 1 CID, so a limit of 0 appears to break. But this is easily worked around by only applying the test to NCID frames. This is easier than special-casing the handling of CID[0] and CID[1]/SPA. In particular, the logic for handling RETIRE_CID and creation of new NCID becomes much simpler.

As I mentioned before, a default of 0 does not work very well, because the default for migration disabled is also 0 -- yet enabling migration without any NCID would be a protocol violation. The default should be at least 2 (initial and SPA), and I would prefer 3.

>   If this parameter is absent, a default of 8 is assumed.

Reviewer 5: I'd prefer a default of 0.

Reviewer 10: I'm not sure I see a default of 0 as being better than 8. 
If we're going to introduce a TP, and if everyone is going to set the TP all the time, then the default doesn't matter. If not, then I think we want to maintain 8 (or some similar value) for the same reasons that we had it before there was a TP.

Reviewer: The point is that otherwise everyone is going to change it from 8. Using 0 avoids the discussion of who's default is better and is in line with other 0 defaults such as max streams.

Reviewer 5: Agreed, we've consistently gone with 0 being a default whenever possible, even in cases when 0 makes the transport basically useless.

Reviewer 10: Then that works for me!

Author: Done.

Reviewer 9: I we have default = 0, do we also need a "migration not supported" TP? It would seem that not setting a non null value would effectively prevent migration.

Reviewer 2: It would, but you can support CID rotation *without* supporting migration, so the reverse isn't true.

a preferred address.  A server MUST treat receipt of any of these transport
parameters as a connection error of type TRANSPORT_PARAMETER_ERROR.

max_connection_ids (0x000e):

: The maximum number of connection IDs that the peer is willing to store.
  This value includes only connection IDs sent in NEW_CONNECTION_ID frames.

>   If this parameter is absent, a default of 0 is assumed.

Reviewer 4: I'd really like to be able to not send this.  Is 0 really the right answer?

Reviewer 10: We had a recommended value previously, let's pull that value over and have the default value conversation separately?

Author: We discussed this before in https://github.com/quicwg/base-drafts/pull/1998#discussion_r233121475. We're using 0 as the default value for all flow control and stream limit values, so at least this seems consistent.

In Tokyo we had disagreement about what a reasonable default should be. @ianswett advocated for 2 or 3, if I remember correctly, while other people want 8. One reason for moving forward with this PR was exactly the fact that people have different use cases in mind, and there's no one-size-fits-all value.

Reviewer 9: What if an endpoint set both "migration disabled = 0" (default) and "max CID=0" (default)? Should that trigger a protocol error?

Reviewer 9: By the way, we already use no zero defaults for some parameters: ack_delay_exponent defaults to 3, and max_ack_delay defaults to 25. There is no rule that says that forbids us to set a non zero default for the proposed MAX CID parameter. I would think that 3 is a good number.

Reviewer 2: In HTTP, we use zero defaults for everything because those values apply until the SETTINGS frame arrives.  In QUIC, however, you can't send anything until transport parameters are in hand.  Any value can be a reasonable default.  (Note that the requirement not to reduce after accepting 0-RTT will apply here, though.)

Reviewer 4: This is one where reducing after 0-RTT is perfectly OK.  It's not a hard limit, it's more like a guideline.  We should allow an exception here.

Reviewer 2: True -- we have an exemption for the one HTTP setting which is purely advisory.  Since this is a suggestion, reducing it wouldn't necessarily hurt anything.  (But there's no prohibition on the sender to not blow up on exceeding the limit, either.)

Reviewer 7: I am a little hazy on the proper behavior for NAT rebinding vs. migration, but how is communicating a zero here different from disable_migration? Do we still need that TP?

Reviewer 2: Yes, because you might want to rotate CIDs on a connection that isn't actively migrating.  Whether you'll survive NAT rebinding depends on the server infrastructure -- a 2-tuple load balancer will survive a port change, but a 4-tuple load balancer can't guarantee it.  An IP+port change isn't guaranteed to survive on anything but a CID-aware load balancer.

a preferred address.  A server MUST treat receipt of any of these transport
parameters as a connection error of type TRANSPORT_PARAMETER_ERROR.

max_connection_ids (0x000e):

> : The maximum number of connection IDs that the peer is willing to store.

Reviewer 3: ```suggestion
: The maximum number of connection IDs from the peer that an endpoint is willing to store.
```

: A peer has migrated to a different network when the endpoint had disabled
  migration.

CONNECTION_ID_LIMIT_ERROR (0xd):

> : A peer received more connection IDs than its advertised limit.

Reviewer 3: See my comment above

| 0x9   | VERSION_NEGOTIATION_ERROR | Version negotiation failure   | {{error-codes}} |
| 0xA   | PROTOCOL_VIOLATION        | Generic protocol violation    | {{error-codes}} |
| 0xC   | INVALID_MIGRATION         | Violated disabled migration   | {{error-codes}} |

> | 0xD   | CONNECTION_ID_LIMIT_ERROR | Too many connection IDs provided | {{error-codes}} |

Author: The whole table looks like it wants to be reordered at some point.
